File: src/04.stack/08.636.go
package stack

import (
	"strconv"
	"strings"
)

func ExclusiveTime(n int, logs []string) []int {
	res := make([]int, n)
	stack := []int{}
	preTime := 0

	for _, v := range logs {
		vArr := strings.Split(v, ":")
		curTime, _ := strconv.Atoi(vArr[2])
		curIndex, _ := strconv.Atoi(vArr[0])

		if vArr[1] == "start" {
			if len(stack) != 0 {
				res[stack[len(stack)-1]] += curTime - preTime
			}
			stack = append(stack, curIndex)
			preTime = curTime
		} else {
			res[curIndex] += curTime - preTime + 1
			stack = stack[:len(stack)-1]
			preTime = curTime + 1
		}
	}

	return res
}

File: src/04.stack/08.636.ts
export function exclusiveTime(n: number, logs: string[]): number[] {
    const res = new Array(n).fill(0);
    const stack: string[] = [];
    let preTime = 0;

    for (let i = 0; i < logs.length; i++) {
        const itemList = logs[i].split(":");
        const curTime = Number(itemList[2]);
        const curIndex = itemList[0];

        if (itemList[1] === "start") {
            if (stack.length) {
                res[stack[stack.length - 1]] += curTime - preTime;
            }
            stack.push(curIndex);
            preTime = curTime;
        } else {
            res[curIndex] += curTime - preTime + 1;
            stack.pop();
            preTime = curTime + 1;
        }
    }

    return res;
}
